src/workerd/api/actor-destroy.c++

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "actor-destroy.h"
+#include <workerd/api/global-scope.h>
+#include <workerd/jsg/ser.h>
+
+namespace workerd::api {
+
+ActorDestroyEvent::ActorDestroyEvent() : ExtendableEvent("actorDestroy"){};
+
+kj::Promise<WorkerInterface::CustomEvent::Result>
+ActorDestroyCustomEventImpl::run(kj::Own<IoContext::IncomingRequest> incomingRequest,
+                                 kj::Maybe<kj::StringPtr> entrypointName) {
+
+  auto& context = incomingRequest->getContext();
+
+  // Mark the request as delivered because we're about to run some JS.
+  incomingRequest->delivered();
+
+  EventOutcome outcome = EventOutcome::OK;
+
+  try {
+    auto result = co_await context.run(
+        [entrypointName = entrypointName, &context](Worker::Lock& lock) mutable {
+          return lock.getGlobalScope().actorDestroy(
+              lock, lock.getExportedHandler(entrypointName, context.getActor()));
+        });
+
+    if (result.outcome == EventOutcome::OK) {
+      // Now that we've completed the handler execution we can do all the necessary cleanup and
+      // cancel prevent anything else to be sent to the destroyed actor. As such, we abort the
+      // current IoContext and shutdown the actor.
+      context.getActorOrThrow().shutdown(
```

Comment:
I think we want to invoke this inside the blockConcurrencyWhile in ServiceWorkerGlobalScope::actorDestroy(). (We could probably be in an immediate continuation instead since kj promises are depth first without yielding.) We want to promise to our users that no additional worker code can run and no more storage api usage will be allowed after their destroy handler runs. That said, I wouldn't be surprised if we end up in some sort of cancellation hell with that. :upside_down_face:

Reply:
You're right and I had thought of that. I'm only afraid we will timeout `blockConcurrencyWhile` which I think has a maximum execution time of 30s. But I guess the context abort and actor destroy won't take any meaningful time, so it should be ok.